### PR TITLE
M4347: Audit log data CRUD events

### PR DESCRIPTION
**Pull request by tommydeboer**
TODO:
- [x] Write tests
- [x] Log SYSTEM actions that were triggered by a user
- [x] Add setting to turn on/off system entity type audit logging
- [x] Exclude some system entity types from audit logging (`IndexAction`s, etc.)
- [x] Log transaction ID and transaction success/failed events

#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- [x] Migration step added in case of breaking change
- [x] User documentation updated
- [x] (If you have changed REST API interface) view-swagger.ftl updated
- [x] Test plan template updated
- [ ] Clean commits
- [ ] Added Feature/Fix to release notes

### CODE REVIEW

File: molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditTransactionListener.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  public AuditTransactionListener(
+      TransactionManager transactionManager, AuditEventPublisher auditEventPublisher) {
+    requireNonNull(transactionManager).addTransactionListener(this);
+    this.auditEventPublisher = requireNonNull(auditEventPublisher);
+  }
+
+  @Override
+  public void rollbackTransaction(String transactionId) {
+    if (isRunByUser()) {
```
Comment (fdlk) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditTransactionListener.java:33]: In the case of jobs, not entirely sure if the context is still present at this moment

File: molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+  @Override
+  public void forEachBatched(Fetch fetch, Consumer<List<Entity>> consumer, int batchSize) {
+    if (isRunByUser() && isNonSystemEntityType()) {
+      delegate()
+          .forEachBatched(
+              fetch,
+              entities -> {
+                consumer.accept(entities);
+                var ids = entities.stream().map(Entity::getIdValue).collect(toList());
+                audit(ENTITIES_READ, ENTITY_IDS, ids);
```
Comment (fdlk) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java:78]: this means the logging happens in batches

File: molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+      audit(ENTITY_DELETED, ENTITY_ID, id);
+    }
+  }
+
+  private void audit(String type) {
+    var data = createDataMap();
+    auditEventPublisher.publish(getUsername(), type, data);
+  }
+
+  private void audit(String type, String k1, Object v1) {
```
Comment (fdlk) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java:293]: key, value probably slightly more meaningful parameter names

Comment (tommydeboer) [reply molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java:293]: Ah yes, I was copying the `Map.of(v1, v2, k1, v2)` style here but I had multiple methods then :)

File: molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java (hunk 2 above)
Comment (tommydeboer) [reply molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java:78]: Correct! :)

File: molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditTransactionListener.java (hunk 1 above)
Comment (tommydeboer) [reply molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditTransactionListener.java:33]: I'll have a go at testing that

Comment (tommydeboer) [reply molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditTransactionListener.java:33]: For jobs a `SystemSecurityToken` is set I believe

Comment (fdlk) [reply molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditTransactionListener.java:33]: Depends on the job, when a job is run as a user, a token is set for the duration of the transaction. What I wonder is when exactly the token gets unset. Before or after the rollbackTransaction is called

Comment (tommydeboer) [reply molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditTransactionListener.java:33]:
```{"timestamp":"2021-03-04T07:56:30.058490Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht36rlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.086132Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4ahlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.110494Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4bxlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.177179Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaae","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.177521Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaai","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.177796Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaam","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.178157Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaaq","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.178516Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaau","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.178880Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaay","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.203098Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4hplg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.240295Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"sys_FileMeta","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4iflg6jaabciqabe"}}
{"timestamp":"2021-03-04T07:56:30.284701Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4mvlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.294393Z","principal":"admin","type":"TRANSACTION_FAILURE","data":{"transactionId":"aaaac574ht4mvlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.308167Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4nvlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.412463Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4uplg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:42.509420Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_ScheduledJob","entityId":"aaaac574gu2mplg6jaabciqaai","transactionId":"aaaac574huuh7lg6jaabciqaae"}}
```

Triggered by the scheduler. Looking good!

File: molgenis-data-platform/src/main/java/org/molgenis/data/platform/decorators/AuditingRepositoryDecoratorFactory.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ *   <li>The repository is not excluded from auditing
+ * </ul>
+ *
+ * For data entity types:
+ *
+ * <ul>
+ *   <li>Auditing of data entity types is set to ALL in the {@link AuditSettings}, OR
+ *   <li>Auditing of data entity types is set to TAGGED in the {@link AuditSettings} and the entity
+ *       type is tagged with the 'audit-usage' tag.
+ * </ul>
```
Comment (fdlk) [inline molgenis-data-platform/src/main/java/org/molgenis/data/platform/decorators/AuditingRepositoryDecoratorFactory.java:41]: I'd put the evaluation of these criteria in a method, that's actually easier to read then

File: molgenis-data/src/main/java/org/molgenis/data/decorator/DynamicRepositoryDecoratorRegistryImpl.java [hunk 5]
```diff
@@ -125,9 +118,4 @@ public DynamicRepositoryDecoratorFactory getFactory(String id) {
       return emptyMap();
     }
   }
-
-  @EventListener
```
Comment (fdlk) [inline molgenis-data/src/main/java/org/molgenis/data/decorator/DynamicRepositoryDecoratorRegistryImpl.java:129]: Not entirely sure if it's an improvement to move the listening to the molgenis repository decorator factory. I kind of liked how the logic of whether the repository should be decorated or not was kept out of the Molgenis decorator factory and each decorator took care of its own logic.

**Review by fdlk — CHANGES_REQUESTED** [review]
Specs changed a bit, discussed with @tommydeboer that we'll await those changes before merging. Two comments for your consideration, am not completely sure about either myself.

**Review by fdlk — APPROVED** [review]

**Pull request merged**